tool/tsh/common/scan.go

```diff
@@ -0,0 +1,357 @@
+	for _, key := range keys {
+		keysSlice = append(keysSlice, key)
+	}
+
+	if err := stream.Send(&accessgraphsecretsv1pb.ReportSecretsRequest{
+		Payload: &accessgraphsecretsv1pb.ReportSecretsRequest_PrivateKeys{
+			PrivateKeys: &accessgraphsecretsv1pb.ReportPrivateKeys{
+				Keys: keysSlice,
+			},
+		},
+	}); err != nil {
+		return trace.Wrap(err, "failed to send private keys")
+	}
+
+	if err := stream.CloseSend(); err != nil {
+		return trace.Wrap(err, "failed to close send")
+	}
+
+	if _, err := stream.Recv(); err != nil && !errors.Is(err, io.EOF) {
+		return trace.Wrap(err, "error closing the stream")
+	}
+
+	fmt.Fprintf(os.Stdout, "Reported %d SSH fingerprints to Teleport.\n", len(keys))
+
+	return nil
+}
+
+func (c *scanKeysCommand) scan(ctx context.Context, deviceID string) map[string]*accessgraphsecretsv1pb.PrivateKey {
+	allKeys := make(map[string]*accessgraphsecretsv1pb.PrivateKey)
+	for _, dir := range c.dirs {
+		keys := walkAndCheckFiles(ctx, dir, deviceID)
+		maps.Copy(allKeys, keys)
+	}
+
+	return allKeys
+}
+
+func (c *scanKeysCommand) print(out io.Writer, keys map[string]*accessgraphsecretsv1pb.PrivateKey) {
+	if len(keys) == 0 {
+		fmt.Fprintf(out, "No SSH private keys found.\n")
+		return
+	}
+	publicKeyMode := func(mode accessgraphsecretsv1pb.PublicKeyMode) string {
+		switch mode {
+		case accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_PUB_FILE:
+			return "used public key file"
+		case accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_PROTECTED:
+			return "protected private key"
+		case accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_DERIVED:
+			return "derived from private key"
+		default:
+			return "unknown"
+		}
+
+	}
+	fmt.Fprintf(out, "%d SSH private keys found.\n", len(keys))
+	for path, key := range keys {
+		fmt.Fprintf(out, "- SHA256: %q (%s) at %s\n",
+			key.Spec.PublicKeyFingerprint,
+			publicKeyMode(key.Spec.PublicKeyMode),
+			path,
+		)
+	}
+}
+
+var (
+	supportedPrivateKeysHeaders = [][]byte{
+		[]byte("RSA PRIVATE KEY"),
+		[]byte("PRIVATE KEY"),
+		[]byte("EC PRIVATE KEY"),
+		[]byte("DSA PRIVATE KEY"),
+		[]byte("OPENSSH PRIVATE KEY"),
+	}
+)
+
+// isSSHPrivateKey checks if a file is an OpenSSH private key
+func isSSHPrivateKey(filePath string) (bool, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return false, err
+	}
+	defer file.Close()
+
+	// read the first 60 bytes of the file
+	var buf [150]byte
+	n, err := file.Read(buf[:])
+	if errors.Is(err, io.EOF) || n < len(buf) {
+		return false, nil
+	} else if err != nil {
+		return false, trace.Wrap(err, "failed to read file")
+	}
+
+	for _, header := range supportedPrivateKeysHeaders {
+		if bytes.Contains(buf[:], header) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// walkAndCheckFiles walks through all files in a directory and its subdirectories
+// and checks if they are OpenSSH private keys
+func walkAndCheckFiles(ctx context.Context, root, deviceID string) map[string]*accessgraphsecretsv1pb.PrivateKey {
```

Comment:
nit: the function not just checks, but returns information collected about matching files. Naming could be improved: `findPrivateKeys` or similar.